lib/tictactoe/console_io.ex

```diff
@@ -0,0 +1,76 @@
+defmodule TicTacToe.ConsoleIO do
+  alias TicTacToe.Board
+  alias TicTacToe.Game
+  alias TicTacToe.Player
+
+  def input, do: String.trim(IO.gets(""))
+
+  def output(message), do: IO.puts(message)
+
+  def greeting, do: output("Welcome to Tic Tac Toe!\n")
+
+  def new_game, do: output("Starting new game...\n")
+
+  def instructions(%Game{} = game) do
+    game
+    |> build_instructions()
+    |> output()
+  end
+
+  def turn_message(%Player{name: name, type: :human}), do: output("#{name}'s turn:")
+
+  def selection_error(reason),
+    do: output("Invalid selection! #{reason} Please try again:")
+
+  def game_won(%Board{} = board, %Player{name: name}) do
+    display_board(board)
+    output("Game over! #{name} wins!")
+  end
+
+  def goodbye, do: output("Thanks for playing! Goodbye.")
+
+  def display_board(%Board{cells: cells}) do
+    length = row_length(cells)
+
+    cells
+    |> Enum.chunk_every(length)
+    |> Enum.map_join(horizontal_line(length), &format_row(&1))
+    |> then(&"#{&1}\n")
+    |> output()
+  end
+
+  defp row_length(cells) do
+    cells
+    |> length()
+    |> :math.sqrt()
+    |> trunc()
+  end
+
+  defp format_row(row_array), do: Enum.map_join(row_array, "|", &format_cell(&1))
+
+  defp format_cell(cell), do: " #{cell} "
+
+  defp horizontal_line(length) do
+    "---"
+    |> List.duplicate(length)
+    |> Enum.join("|")
+    |> then(&"\n#{&1}\n")
+  end
+
+  defp build_instructions(%Game{
+         board: %Board{cells: cells},
+         current_player: current_player,
+         next_player: next_player
```

Comment:
Non-blocking + FYI - if you need to access various keys that have the same name, you could assign them to different values:

```elixir
current_player: %Player{marker: current_player_marker, name: current_player_name},
next_player: %Player{marker: next_player_marker, name: next_player_name}
```